server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java

```diff
@@ -0,0 +1,330 @@
+            } finally {
+                watcherThreadLatch.countDown();
+            }
+        }, "elasticsearch[file-settings-watcher]").start();
+    }
+
+    synchronized void stopWatcher() {
+        logger.debug("stopping watcher ...");
+        if (watching()) {
+            try {
+                if (settingsDirWatchKey != null) {
+                    settingsDirWatchKey.cancel();
+                    settingsDirWatchKey = null;
+                }
+                fileUpdateState = null;
+                watchService.close();
+                if (watcherThreadLatch != null) {
+                    watcherThreadLatch.await();
+                }
+            } catch (IOException | InterruptedException e) {
+                logger.warn("encountered exception while closing watch service", e);
+            } finally {
+                watchService = null;
+                logger.info("watcher service stopped");
+            }
+        } else {
+            logger.debug("file settings service already stopped");
+        }
+    }
+
+    private WatchKey enableSettingsWatcher(WatchKey previousKey, Path settingsDir) throws IOException {
+        if (previousKey != null) {
+            previousKey.cancel();
+        }
+        return settingsDir.register(
+            watchService,
+            StandardWatchEventKinds.ENTRY_MODIFY,
+            StandardWatchEventKinds.ENTRY_CREATE,
+            StandardWatchEventKinds.ENTRY_DELETE
+        );
+    }
+
+    void processFileSettings(Path path, boolean onStartup) throws IOException {
+        logger.info("processing path [{}] for [{}]", path, NAMESPACE);
+        try (var json = new BufferedInputStream(Files.newInputStream(path))) {
+            try (XContentParser parser = JSON.xContent().createParser(XContentParserConfiguration.EMPTY, json)) {
+                stateService.process(NAMESPACE, parser, (e) -> {
+                    if (e != null) {
+                        // If we encountered an exception trying to apply the operator state at
+                        // startup time, we throw an error to force Elasticsearch to exit.
+                        if (onStartup) {
+                            throw new OperatorConfigurationError("Error applying operator settings", e);
+                        } else {
+                            logger.error("Error processing operator settings json file", e);
+                        }
+                    }
+                });
+            }
+        }
+    }
+
+    /**
+     * Holds information about the last known state of the file we watched. We use this
+     * class to determine if a file has been changed.
+     */
+    record FileUpdateState(long timestamp, String path, Object fileKey) {}
+
+    /**
+     * Error subclass that is thrown when we encounter a fatal error while applying
+     * the operator cluster state at Elasticsearch boot time.
+     */
+    public static class OperatorConfigurationError extends Error {
```

Comment:
We shouldn't be extending Error (and should be renamed to avoid configuration, perhaps mentioning this is a startup problem in the name). This can extend RuntimeException so it can be propagated easily.